test-apps/display-test-app/src/frontend/MacroTools.ts

```diff
@@ -0,0 +1,50 @@
+/*---------------------------------------------------------------------------------------------
+* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+* See LICENSE.md in the project root for license terms and full copyright notice.
+*--------------------------------------------------------------------------------------------*/
+import { IModelApp, MessageBoxIconType, MessageBoxType, NotifyMessageDetails, OutputMessagePriority, ParseAndRunResult, Tool } from "@itwin/core-frontend";
+import { DtaRpcInterface } from "../common/DtaRpcInterface";
+
+export class MacroTool extends Tool {
+  public static override toolId = "Macro";
+  public static override get minArgs() { return 1; }
+  public static override get maxArgs() { return 1; }
+
+  public override async run(macroFile: string): Promise<boolean> {
+    const macroString = await DtaRpcInterface.getClient().readExternalFile(macroFile);
+    const commands = macroString.split ("\r\n");
+    commands.forEach((item, index) => {
+      if(item === "") commands.splice(index,1);
+    });
+    if (commands.length === 0)
+      IModelApp.notifications.outputMessage(new NotifyMessageDetails(OutputMessagePriority.Info, "File not found or no content"));
+    else {
+      for (const cmd of commands) {
+        let message: string | undefined;
+        try {
+          switch (await IModelApp.tools.parseAndRun(cmd)) {
+            case ParseAndRunResult.ToolNotFound:
+              message = `Cannot find a key-in that matches: ${cmd}`;
+              break;
+            case ParseAndRunResult.BadArgumentCount:
+              message = "Incorrect number of arguments";
```

Comment:
Here and below include the keyin that failed in the message so it will be easier to fix the script.

Reply:
Good idea, done